Feature/documentation (#23)
orbit-love/member-directory: README.md

5. [General tips](#general-tips)
   1. [Changing the DB structure](#changing-the-db-structure)

## What is it?

This app lets you host a subset of the members in your community and let them share their social profiles & a short bio about themselves. It is ideal for promoting engagement and fostering relationships between members.

![The landing page of the member directory, with Orbit branding.](https://github.com/orbit-love/member-directory/assets/45462299/a1bad632-4437-4aaf-9459-e310e17c04c2)

## Technical Overview

Diagram here...

### Data

[Prisma](https://www.prisma.io/) is used to interface with a Postgres database that exists between the Orbit API and the member directory. This provides control over the data, allowing for manipulation independent of your main data store n Orbit. For example, we've added "bio" and "admin" fields not required by the Orbit application.

Data is fetched from Orbit using the /api/populate-members-table API route. This process fetches members fitting specific criteria (presence of an email and a defined tag in environment variables) and upserts them into Prisma. Members that no longer meet these criteria and still exist in Prisma are removed.

> 

Reviewer: Some suggestions
1. Use simpler words (fitting, upserts)
2. Reduce complexity (This process fetches members → we fetch members)

You can review the database schema at prisma/schema.prisma.

helpers/prisma-helpers.js is used for managing interactions with Prisma.

### Authentication

> We employ [next-auth](https://next-auth.js.org/) for passwordless authentication management in the directory.

Reviewer: ```suggestion
We use [next-auth](https://next-auth.js.org/) for passwordless authentication management in the directory.
```

- Customised pages are located under `pages/auth`.
- The custom email template is stored under `helpers/next-auth-helpers.js`.
- Configuration settings are found in `pages/api/auth/[...nextauth].js`. Some significant configuration changes are listed here:

> 1. Upon a user's sign-in attempt, we first verify their existence in the directory database. This means only listed directory members can access it.

Reviewer: ```suggestion
1. When users try to sign-in, we first verify their existence in the directory database. This means only listed directory members can access it.
```

```
# pages/auth/error.js

if (error === "AccessDenied") {
router.replace("/auth/verify-request");
}
```

3. Once a user signs in, we set the "admin" flag in the session object based on their admin status in the database. This allows us to modify the UI based on whether a member is an admin or not.

#### Securing Frontend Pages

The app has two layouts: LayoutAuthenticated and LayoutUnauthenticated. All pages utilize one of these layouts. Both have a hook that runs on load, fetching the current session and performing the appropriate redirection:

- Unauthenticated users attempting to access a protected page are redirected to the login page.
- Authenticated users trying to access an unauthenticated page are redirected to the full directory.

#### Securing API Endpoints

Protection for API routes is achieved with higher-order functions defined in `helpers/api-helpers.js`. One such function, `withAuthCheck`, ensures a user is signed in before granting them access to the endpoint. For more details on these, refer to the next section, [API Protection](#api-protection).

> ### API protection

Reviewer: IMHO we can remove all the parts related to RoR. Most developers don't need to know about it.

This should be all you need to run the app and receive emails, but you still won't be able to sign in without any data. So...

### Initial data

There is a convenient API route for initialising your first admin user.

Send a POST request to `localhost:3000/api/initialise-admin-member`. If it responds with a 200 OK & a created member object, you're all set to log in & move on to the next step.

### Running the server, logging in

Right, I feel like that was a lot 😪 But don't worry, that's all one-time setup. From now-on you should be clear to just:

```
yarn dev
```

& visit [localhost:3000](http://localhost:3000) to see the directory, and [localhost:8025](http://localhost:8025) to see incoming emails to sign in.

And it's

```
yarn test --watch
```

to run the automated tests in an instance where they hot-reload with changes.

As an admin you should be able to sync members from the UI once you've signed in, which will pull the latest data from Orbit.

## Setting up your own instance

### Branding

We aim to keep this app flexible, and require minimal changes to switch over the branding.

#### Update brand colours

> You can update the brand colors in the `tailwind.config.js` file by modifying these variables:

Reviewer: Maybe you can add the direct link to the file. (e.g https://github.com/orbit-love/member-directory/blob/main/tailwind.config.js#L30)

```
accent: "#6C4DF6"
"accent-highlight": "#5B41CF"
dark: "#1E2124"
"dark-highlight": "#3A4045"
light: "#FFFFFF"
"light-highlight": "#D3D6DA"
```

Each color variable is documented in the same file for ease of reference.

**Important** You will also need to copy-paste the "accent" value into the brandColor variable in helpers/next-auth-helpers.js. This will be used to colour the button in the email templates.

### Update brand images

Replace the following three images to the public/ directory:

> 1. `brand-icon-dark.svg`, which will be shown in the header bar on light theme screens

Reviewer: Same here, add a link to the file.
